### DISCUSSION TITLE
Add delay between requests durning collection run

### DISCUSSION THREAD

**Discussion by buont** [usebruno/bruno#420] [Ideas]
Hi,

I think a very welcome feature is to add the ability to define a delay between requests in a sequence when running a collection.

Thanks

**Comment by martinsefcik** [comment]
Until it is supported you can add delay manually in requests where it is needed by putting this in Post response script
```js
await new Promise(r => setTimeout(r, 1500));
```

**Reply by helloanoop** [reply]
Nice. You can do this for all requests by setting it at the collection script level.

**Reply by LesJW** [reply]
> Until it is supported you can add delay manually in requests where it is needed by putting this in Post response script
> 
> ```js
> await new Promise(r => setTimeout(r, 1500));
> ```

Hello martinsefcik,

I paste this line into the Post response script, but it shows an error with the word "new" underlined. The error is: "Expected an operator and instead saw 'new'"

I don't know Javascript, but after researching and trial and error, I came up with the following code, which does not show any error in Bruno. The problem is, it doesn't actually insert the delay when the collection is ran. I'm not sure how to debug.

function resolveAfter2Seconds() {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve('calling');
    }, 2000);
  });
}

async function asyncCall() {
  var result = await resolveAfter2Seconds();
  console.log(result);
  // expected output: "resolved"
  console.log('called');
}

asyncCall();

Any advice on how to get your solution in the response script without error? I'm running on Windows 11.

Thanks.

**Reply by Its-treason** [reply]
Hey @LesJW,

eventhrough the code editor shows an error, this snippet still works: `await new Promise(r => setTimeout(r, 1500));`. The Code editor sometimes does not understand newer Javascript.

About your code snippet:

```diff
  function resolveAfter2Seconds() {
  return new Promise(resolve => {
  setTimeout(() => {
  resolve('calling');
  }, 2000);
  });
  }

  async function asyncCall() {
  var result = await resolveAfter2Seconds();
  console.log(result);
  // expected output: "resolved"
  console.log('called');
  }

- asyncCall();
+ await asyncCall();
```
You have to add `await` when calling `asyncCall`. 

I would recommeend learning more about Promises and async/await. This site seems to have a good tutorial: https://javascript.info/async

**Reply by LesJW** [reply]
Thanks for replying! Yes, the snippet does run without producing errors. But the delay between requests is still zero. I changed it to 10 seconds, and noticed that after running all iterations (20 in my case), it delays for 10 sec before completing. So it's putting the delay after all the iterations have ran!

I then removed the snippet from the request and put it at the collection level. Same thing, runs all requests, then delays.

Regarding my code with the functions, I added the "await" as you suggested, and it runs as well. It also puts the delay after all requests have completed. So it appears both the snippet and my code are good, but Bruno is putting the delay at the end, instead of between requests. If I put the script in the "Pre-Request" section, it waits 10 seconds, then runs all the requests at once. Not sure where to go next. Thanks.

**Reply by Its-treason** [reply]
> I changed it to 10 seconds, and noticed that after running all iterations (20 in my case), it delays for 10 sec before completing. So it's putting the delay after all the iterations have ran!

I see, you used the "Iterations" options and you want a delay between the iterations. So instead of running parralell they should run in sequence? There is currenlty an open issue for this: https://github.com/usebruno/bruno/issues/2251 Its not possible to achive with scripting.

**Comment by LesJW** [comment]
Ahh, thank you. Now I understand. I came from Postman that when iterating through a CSV file on a request, you can specify a time delay between each API call. I thought this feature in Bruno was the equivalent.

I see the open issue 2251. I would suggest also adding a time delay between calls if the parallel checkbox is unselected. I call several different API endpoints to update thousands of records. Some of them have restrictions on the number of calls - like you can't exceed 100 calls in any 10 second period. Even with parallel turned off, you could exceed such a condition, depending on how fast it responds to each call.

Thanks for the info and clarification. I'll try to figure out another way to do my mass updates.

**Comment by TheMightyGit** [comment]
```js
await new Promise(r => setTimeout(r, 2000));
```

...gives me the 'new' error in all forms of running (bruno 1.22.0 macos), but this seems to work...

```js
function msleep(n) {
  Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, n);
}

msleep(2000); // 2 second sleep
```

(as seen here - https://stackoverflow.com/a/74210440 )

**Reply by orgads** [reply]
You can just enclose the new Promise in parentheses:
```js
await (new Promise(r => setTimeout(r, 1500)));
```
